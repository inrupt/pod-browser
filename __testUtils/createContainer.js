/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addUrl,
  createSolidDataset,
  createThing,
  addDatetime,
  addDecimal,
  addInteger,
  setThing,
} from "@inrupt/solid-client";
import { rdf, ldp } from "rdf-namespaces";

export const TIMESTAMP = new Date(Date.UTC(2020, 5, 2, 15, 59, 21));

export default function createContainer(
  url = "https://user.dev.inrupt.net/",
  type = ldp.BasicContainer
) {
  const thing = createThing({ url });
  const publicContainer = [
    (t) => addUrl(t, rdf.type, type),
    (t) => addUrl(t, rdf.type, ldp.Container),
    (t) => addDatetime(t, "http://purl.org/dc/terms/modified", TIMESTAMP),
    (t) =>
      addDecimal(t, "http://www.w3.org/ns/posix/stat#mtime", 1591131561.195),
    (t) => addInteger(t, "http://www.w3.org/ns/posix/stat#size", 4096),
    (t) => addUrl(t, ldp.contains, "https://user.dev.inrupt.net/public/games/"),
  ].reduce((acc, fn) => fn(acc), thing);

  return setThing(createSolidDataset(), publicContainer);
}
