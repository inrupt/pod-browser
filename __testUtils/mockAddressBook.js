/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addStringNoLocale,
  addUrl,
  createSolidDataset,
  mockSolidDatasetFrom,
  mockThingFrom,
  setThing,
} from "@inrupt/solid-client";
import { acl, dc, rdf } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { aliceWebIdUrl } from "./mockPersonResource";
import { vcardExtras } from "../src/addressBook";
import { getAddressBookIndexDefaultUrl } from "../src/models/addressBook";
import { joinPath } from "../src/stringHelpers";

export const defaultContainerUrl = "https://user.example.com/contacts";
export const defaultAddressBookDatasetUrl = joinPath(
  defaultContainerUrl,
  "index.ttl"
);
export const defaultAddressBookThingUrl = `${defaultAddressBookDatasetUrl}#this`;

export function mockAddressBookThing(options = {}) {
  const containerUrl = options.containerUrl || defaultContainerUrl;
  const owner = options.owner || aliceWebIdUrl;
  const indexUrl =
    options.indexUrl || getAddressBookIndexDefaultUrl(containerUrl);
  const mainIndexUrl = `${indexUrl}#this`;
  return chain(
    mockThingFrom(mainIndexUrl),
    (t) => addUrl(t, rdf.type, vcardExtras("AddressBook")),
    (t) => addUrl(t, acl.owner, owner),
    (t) => addStringNoLocale(t, dc.title, "Contacts")
  );
}

export function mockAddressBookDataset(
  options = {},
  thing = mockAddressBookThing(options)
) {
  const containerUrl = options.containerUrl || defaultContainerUrl;
  const indexUrl =
    options.indexUrl || getAddressBookIndexDefaultUrl(containerUrl);
  return chain(mockSolidDatasetFrom(indexUrl), (d) => setThing(d, thing));
}

export default function mockAddressBook(options = {}) {
  const containerUrl = options.containerUrl || defaultContainerUrl;
  const thing = mockAddressBookThing(options);
  return {
    containerUrl,
    thing,
    dataset: mockAddressBookDataset(options, thing),
  };
}

export function mockEmptyAddressBook(containerUrl, owner, title = "Contacts") {
  const thing = chain(
    mockAddressBookThing({ containerUrl, owner }),
    (t) => addUrl(t, rdf.type, vcardExtras("AddressBook")),
    (t) => addUrl(t, acl.owner, owner),
    (t) => addStringNoLocale(t, dc.title, title)
  );
  return {
    containerUrl,
    dataset: setThing(createSolidDataset(), thing),
    thing,
  };
}
