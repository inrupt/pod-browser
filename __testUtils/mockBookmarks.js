/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addDatetime,
  addStringNoLocale,
  addUrl,
  mockSolidDatasetFrom,
  mockThingFrom,
  setThing,
} from "@inrupt/solid-client";
import { dct, rdf } from "rdf-namespaces";
import {
  RECALLS_PROPERTY_IRI,
  BOOKMARK_TYPE_IRI,
} from "../src/solidClientHelpers/bookmarks";

const bookmarksUrl = "http://example.com/bookmarks/index.ttl";
const bookmark1Title = "Cat gifs";
const bookmark1Recalls = "https://example.com/catgifs";
const bookmark2Title = "Dog videos";
const bookmark2Recalls = "https://example.com/dogvideos";

export default function mockBookmarks() {
  let bookmarksDataset = mockSolidDatasetFrom(bookmarksUrl);
  const bookmark1 = mockThingFrom(`${bookmarksUrl}#1234`);
  const bookmark1WithTitle = addStringNoLocale(
    bookmark1,
    dct.title,
    bookmark1Title
  );
  const bookmark1WithRecalls = addUrl(
    bookmark1WithTitle,
    RECALLS_PROPERTY_IRI,
    bookmark1Recalls
  );
  const bookmark1WithType = addUrl(
    bookmark1WithRecalls,
    rdf.type,
    BOOKMARK_TYPE_IRI
  );
  const bookmark1WithCreated = addDatetime(
    bookmark1WithType,
    RECALLS_PROPERTY_IRI,
    new Date(Date.UTC(2020, 9, 28, 3, 24, 0))
  );
  bookmarksDataset = setThing(bookmarksDataset, bookmark1WithCreated);
  const bookmark2 = mockThingFrom(`${bookmarksUrl}#4567`);
  const bookmark2WithTitle = addStringNoLocale(
    bookmark2,
    dct.title,
    bookmark2Title
  );
  const bookmark2WithRecalls = addUrl(
    bookmark2WithTitle,
    RECALLS_PROPERTY_IRI,
    bookmark2Recalls
  );
  const bookmark2WithType = addUrl(
    bookmark2WithRecalls,
    rdf.type,
    BOOKMARK_TYPE_IRI
  );
  const bookmark2WithCreated = addDatetime(
    bookmark2WithType,
    RECALLS_PROPERTY_IRI,
    new Date(Date.UTC(2020, 9, 29, 3, 24, 0))
  );
  bookmarksDataset = setThing(bookmarksDataset, bookmark2WithCreated);
  return {
    dataset: bookmarksDataset,
    iri: bookmarksUrl,
  };
}
