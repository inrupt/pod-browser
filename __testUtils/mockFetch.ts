/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import "whatwg-fetch";
import Mock = jest.Mock;

type fetchType = typeof window.fetch;

export default function mockFetch(
  responses: Record<string, Response | Record<string, Response>>
): Mock<fetchType> {
  return jest
    .fn<any, Parameters<fetchType>>()
    .mockImplementation((url, init) => {
      const response = responses[url.toString()];
      const responseObject = response as Record<string, Response>;
      const method = init?.method || "GET";
      if (typeof responseObject[method] !== "undefined") {
        return Promise.resolve(responseObject[method]);
      }
      if (response) {
        return Promise.resolve(response);
      }
      throw new Error(`URL (${url}) not mocked properly`);
    });
}

export function mockSolidClientRequest(
  body: string,
  method = "GET",
  headers: Record<string, string> = {}
): Request {
  return new Request(body, {
    headers: new Headers(headers),
    method,
  });
}

export function mockTurtleResponse(status: number, body = ""): Response {
  return new Response(body, {
    status,
    headers: {
      "Content-Type": "text/turtle",
    },
  });
}
