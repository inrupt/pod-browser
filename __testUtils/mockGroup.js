/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addUrl,
  getStringNoLocale,
  getThing,
  mockSolidDatasetFrom,
  mockThingFrom,
  setStringNoLocale,
  setThing,
  setUrl,
} from "@inrupt/solid-client";
import { rdf, vcard } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { createGroupDatasetUrl } from "../src/models/group";
import { vcardExtras } from "../src/models/addressBook";

function mockGroupThing(name, groupThingUrl) {
  return chain(
    mockThingFrom(groupThingUrl),
    (t) => setUrl(t, rdf.type, vcard.Group),
    (t) => setStringNoLocale(t, vcard.fn, name)
  );
}

function mockIndexThing(addressBook, groupThingUrl) {
  return setUrl(addressBook.thing, vcardExtras("includesGroup"), groupThingUrl);
}

export function addGroupToIndexDataset(
  dataset,
  addressBook,
  name,
  groupThingUrl
) {
  return chain(
    dataset,
    (d) => setThing(d, mockGroupThing(name, groupThingUrl)),
    (d) => setThing(d, mockIndexThing(addressBook, groupThingUrl))
  );
}

// export function addGroupsToAddressBook(addressBook, groupsToAdd) {
//   const { groups, ...rest } = addressBook;
//   return {
//     ...rest,
//     groups: {
//       iri: addressBook.groups.iri,
//       dataset: chain(
//         addressBook.groups.dataset,
//         ...groupsToAdd.map(({ dataset, iri }) => (d) => {
//           const group = getThing(dataset, iri);
//           const name = getStringNoLocale(group, vcard.fn);
//           return setThing(d, mockGroupThing(name, iri));
//         }),
//         ...groupsToAdd.map((group) => (d) =>
//           setThing(d, mockIndexThing(addressBook, group.iri))
//         )
//       ),
//     },
//   };
// }

export default function mockGroup(
  addressBook,
  name,
  { url, id, members } = {}
) {
  const groupDatasetUrl = url || createGroupDatasetUrl(addressBook, id);
  const groupThingUrl = `${groupDatasetUrl}#this`;
  const groupThing = chain(
    mockGroupThing(name, groupThingUrl),
    ...(members || []).map((agentUrl) => (t) =>
      addUrl(t, vcard.hasMember, agentUrl)
    )
  );
  return {
    dataset: chain(
      mockSolidDatasetFrom(groupDatasetUrl),
      (d) => setThing(d, groupThing),
      (d) => setThing(d, mockIndexThing(addressBook, groupThingUrl))
    ),
    thing: groupThing,
  };
}
