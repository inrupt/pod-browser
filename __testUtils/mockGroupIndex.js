/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { mockSolidDatasetFrom } from "@inrupt/solid-client";
import { getContactIndexDefaultUrl } from "../src/models/contact/collection";
import { addGroupToMockedIndexDataset } from "./mockGroupContact";
import { getGroupName, getGroupUrl } from "../src/models/group";
import { chain } from "../src/solidClientHelpers/utils";
import { GROUP_CONTACT } from "../src/models/contact/group";

export default function mockGroupIndex(addressBook, groups) {
  const contactIndexUrl = getContactIndexDefaultUrl(
    addressBook.containerUrl,
    GROUP_CONTACT
  );
  return {
    dataset: chain(
      mockSolidDatasetFrom(contactIndexUrl),
      ...groups.map((g) => (d) =>
        addGroupToMockedIndexDataset(
          d,
          addressBook,
          getGroupName(g),
          getGroupUrl(g)
        )
      )
    ),
    type: GROUP_CONTACT,
  };
}
