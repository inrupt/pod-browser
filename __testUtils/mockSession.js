/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import mockFetch from "./mockFetch";

import profile from "./mocks/profile.ttl";
import storageAclTtl from "./mocks/storageAcl.ttl";
import mockResponse from "./mockResponse";

export const webId = "http://example.com/webId#me";
export const storage = "http://example.com/";
export const storageAcl = "http://example.com/.acl";

export { profile };

export default function mockSession(options = {}) {
  // this instance is a normal, authenticated session
  // you can override the properties with what you need, but if you repeat something a lot
  // consider naming it and create a reusable mock function
  return {
    fetch: mockFetch({
      [webId]: () => mockResponse(200, profile),
      [storage]: () =>
        mockResponse(200, "", {
          Link: `<${storageAcl}>; rel="acl"`,
          url: storage,
        }),
      [storageAcl]: () =>
        mockResponse(200, storageAclTtl, {
          url: storageAcl,
        }),
    }),
    info: {
      isLoggedIn: true,
      sessionId: "some-session-id",
      webId,
    }, // add more properties as needed
    ...options,
  }; // add more properties as needed
}

// create other functions in this file to reflect other types of sessions we need to test with

export function mockUnauthenticatedSession() {
  return {
    fetch: () => mockFetch(),
    info: {
      isLoggedIn: false,
      sessionId: null,
      webId: null,
    }, // add more properties as needed
  }; // add more properties as needed
}

export function mockAuthenticatedSessionWithNoAccessToPod() {
  return {
    fetch: mockFetch({
      [webId]: () => mockResponse(200, profile),
      [storage]: () =>
        mockResponse(200, "", {
          Link: `<${storageAcl}>; rel="acl"`,
          url: storage,
        }),
      [storageAcl]: () =>
        mockResponse(401, "", {
          url: storageAcl,
        }),
    }),
    info: {
      isLoggedIn: true,
      sessionId: "some-session-id",
      webId,
    },
  };
}
