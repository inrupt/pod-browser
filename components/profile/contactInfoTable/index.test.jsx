/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { rdf, vcard } from "rdf-namespaces";
import * as scFns from "@inrupt/solid-client";
import { ThingProvider } from "@inrupt/solid-ui-react";

import { addUrl, getThingAll, getUrl, setThing } from "@inrupt/solid-client";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import mockSession from "../../../__testUtils/mockSession";
import mockSessionContextProvider from "../../../__testUtils/mockSessionContextProvider";
import { renderWithTheme } from "../../../__testUtils/withTheme";

import ContactInfoTable, {
  DEFAULT_CONTACT_TYPE,
  PREFIX_MAP,
  setupAddContactDetail,
  setupDeleteButtonCell,
  setupRemoveRow,
  setupSaveHandler,
} from "./index";

const { mockSolidDatasetFrom, mockThingFrom } = scFns;
const dataset = mockSolidDatasetFrom("http://example.com/dataset");
const profile = mockThingFrom("http://example.com/profile#this");

describe("ContactInfoTable", () => {
  test("renders a table of contact info", () => {
    const session = mockSession({ fetch });
    const SessionProvider = mockSessionContextProvider(session);
    const thing = mockThingFrom("http://example.com/alice#me");

    const { asFragment } = renderWithTheme(
      <SessionProvider>
        <ThingProvider thing={thing}>
          <ContactInfoTable property={vcard.hasEmail} />
        </ThingProvider>
      </SessionProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("renders an editable table of contact info", () => {
    const session = mockSession({ fetch });
    const SessionProvider = mockSessionContextProvider(session);
    const thing = mockThingFrom("http://example.com/alice#me");

    const { asFragment } = renderWithTheme(
      <SessionProvider>
        <ThingProvider thing={thing}>
          <ContactInfoTable property={vcard.hasEmail} editing />
        </ThingProvider>
      </SessionProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});

describe("setupAddContactDetail", () => {
  it("handles adding telephone numbers", async () => {
    const property = vcard.hasTelephone;
    const contactType = DEFAULT_CONTACT_TYPE;
    const contactValue = "42-1337";
    const saveHandler = jest.fn();
    const setNewContactValue = jest.fn();

    await expect(
      setupAddContactDetail(
        property,
        contactType,
        contactValue,
        dataset,
        profile,
        saveHandler,
        setNewContactValue
      )()
    ).resolves.toBeUndefined();

    expect(saveHandler).toHaveBeenCalled();
    const [newProfile, newDataset] = saveHandler.mock.calls[0];
    const newContactDetail = getThingAll(newDataset)[0];
    expect(getUrl(newContactDetail, rdf.type)).toEqual(contactType);
    expect(getUrl(newContactDetail, vcard.value)).toEqual(
      `${PREFIX_MAP[property]}${contactValue}`
    );

    expect(setNewContactValue).toHaveBeenCalledWith("");
  });
});

describe("setupSaveHandler", () => {
  it("sets a save handler using saveSolidDatasetAt", async () => {
    const savedDataset = "savedDataset";
    jest.spyOn(scFns, "saveSolidDatasetAt").mockResolvedValue(savedDataset);
    const fetch = jest.fn();
    const setDataset = jest.fn();
    const newThing = mockThingFrom("http://example.com/#newThing");
    const datasetUrl = "http://example.com";
    const datasetToUpdate = mockSolidDatasetFrom(datasetUrl);

    await expect(
      setupSaveHandler(fetch, setDataset)(newThing, datasetToUpdate)
    ).resolves.toBeUndefined();
    expect(scFns.saveSolidDatasetAt).toHaveBeenCalledWith(
      datasetUrl,
      setThing(datasetToUpdate, newThing),
      { fetch }
    );
    expect(setDataset).toHaveBeenCalledWith(savedDataset);
  });
});

describe("setupRemoveRow", () => {
  it("removes details and saves change", async () => {
    const property = vcard.hasTelephone;
    const contactValue = "tel:test";
    const profileWithNumber = addUrl(profile, property, contactValue);
    const saveHandler = jest.fn();
    await expect(
      setupRemoveRow(
        profileWithNumber,
        property,
        saveHandler,
        dataset
      )(mockThingFrom(contactValue))
    ).resolves.toBeUndefined();
    expect(saveHandler).toHaveBeenCalled();
    const [newProfile] = saveHandler.mock.calls[0];
    expect(getUrl(newProfile, property)).toBeNull();
  });
});

describe("setupDeleteButtonCell", () => {
  const thing = mockThingFrom("http://example.com#thing");

  it("returns a button if editable", () => {
    const removeRow = jest.fn();
    const bem = jest.fn();
    const Button = setupDeleteButtonCell(true, removeRow, bem);

    const { asFragment, container } = render(
      <ThingProvider thing={thing}>
        <Button />
      </ThingProvider>
    );

    expect(asFragment()).toMatchSnapshot();
    expect(bem).toHaveBeenCalledWith("button", "action");

    userEvent.click(container.querySelector("button"));
    expect(removeRow).toHaveBeenCalledWith(thing);
  });

  it("renders nothing if not editable", () => {
    const Component = setupDeleteButtonCell(
      false,
      () => {},
      () => {}
    );

    const { asFragment } = render(
      <ThingProvider thing={thing}>
        <Component />
      </ThingProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
