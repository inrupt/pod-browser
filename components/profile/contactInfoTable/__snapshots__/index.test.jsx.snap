// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactInfoTable renders a table of contact info 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-root PodBrowser-root"
  >
    <table
      class="PodBrowser-table  PodBrowser-table--inherits PodBrowser-table"
      role="table"
    >
      <thead>
        <tr
          role="row"
        >
          <th
            colspan="1"
            role="columnheader"
          />
          <th
            colspan="1"
            role="columnheader"
          />
          <th
            colspan="1"
            role="columnheader"
          />
        </tr>
      </thead>
      <tbody
        role="rowgroup"
      />
    </table>
  </div>
</DocumentFragment>
`;

exports[`ContactInfoTable renders an editable table of contact info 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-root PodBrowser-root"
  >
    <table
      class="PodBrowser-table  PodBrowser-table--inherits PodBrowser-table"
      role="table"
    >
      <thead>
        <tr
          role="row"
        >
          <th
            colspan="1"
            role="columnheader"
          />
          <th
            colspan="1"
            role="columnheader"
          />
          <th
            colspan="1"
            role="columnheader"
          />
        </tr>
      </thead>
      <tbody
        role="rowgroup"
      />
    </table>
  </div>
  <div
    class="PodBrowser-root PodBrowser-root"
  >
    <div
      class="PodBrowser-root PodBrowser-root"
    >
      <div
        class="PodBrowser-root"
      >
        <div
          class="PodBrowser-root PodBrowser-root PodBrowser-formControl"
          data-testid="profile-undefined-new-type"
        >
          <div
            aria-haspopup="listbox"
            class="PodBrowser-root PodBrowser-select PodBrowser-selectMenu PodBrowser-outlined PodBrowser-input PodBrowser-input"
            role="button"
            tabindex="0"
          >
            Home
          </div>
          <input
            aria-hidden="true"
            class="PodBrowser-nativeInput"
            tabindex="-1"
            value="http://www.w3.org/2006/vcard/ns#Home"
          />
          <svg
            aria-hidden="true"
            class="PodBrowser-root PodBrowser-icon PodBrowser-iconOutlined"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="PodBrowser-root PodBrowser-notchedOutline"
            style="padding-left: 8px;"
          >
            <legend
              class="PodBrowser-legend"
              style="width: 0.01px;"
            >
              <span>
                â€‹
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="PodBrowser-root PodBrowser-root"
    >
      <input
        class="PodBrowser-input"
        data-testid="profile-undefined-new-value"
        label="Value"
        value=""
      />
    </div>
    <div
      class="PodBrowser-root PodBrowser-root"
    >
      <button
        class="PodBrowser-button PodBrowser-button--action"
        data-testid="profile-undefined-add-button"
        type="button"
      >
        Add
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`setupColumnTypeBody has a fallback if value is not valid 1`] = `
<DocumentFragment>
  <p
    class="MuiTypography-root MuiTypography-body1"
    data-testid="profile-undefined-existing-type"
    title="invalid"
  >
    invalid
  </p>
</DocumentFragment>
`;

exports[`setupColumnTypeBody sets up body for type 1`] = `
<DocumentFragment>
  <p
    class="MuiTypography-root MuiTypography-body1"
    data-testid="profile-undefined-existing-type"
    title="http://www.w3.org/2006/vcard/ns#Home"
  >
    Home
  </p>
</DocumentFragment>
`;

exports[`setupColumnValueBody renders for editing mode 1`] = `
<DocumentFragment>
  <p
    class="MuiTypography-root MuiTypography-body1"
  >
    <span>
      fetching data in progress
    </span>
  </p>
</DocumentFragment>
`;

exports[`setupColumnValueBody renders for read-only mode 1`] = `
<DocumentFragment>
  <p
    class="MuiTypography-root MuiTypography-body1"
  >
    <span>
      fetching data in progress
    </span>
  </p>
</DocumentFragment>
`;

exports[`setupDeleteButtonCell renders nothing if not editable 1`] = `<DocumentFragment />`;

exports[`setupDeleteButtonCell returns a button if editable 1`] = `
<DocumentFragment>
  <button
    data-testid="profile-email-delete-button"
    type="button"
  >
    Delete
  </button>
</DocumentFragment>
`;
